File: Windows/Source/Installer.c
// Basic headers for C files
#include <stdio.h>
#include <stdbool.h>
// Getting the windows header for Internet and System to make internet connections and create the file(s) needed
#include <Windows.h>
#include <wininet.h>
// Linking wininet.lib to this file, so it can use the syscalls without needing to use the Windows API
#pragma comment(lib, "wininet.lib")

int DownloadFromWeb(char *url, char* outputFile)
{
    HINTERNET handleInternet, handleConnection;
    DWORD bytesRead;
    FILE *fptr;
    // Initalizing the internet connection, by allowing this application to connect through the default connected internet
    handleInternet = InternetOpen(TEXT("Downloader"), INTERNET_OPEN_TYPE_DIRECT, NULL, NULL, 0);
    if (handleInternet == NULL)
    {
        printf("[!] ERROR 02: Problem encountered when creating internet connection.\n[#] Rerun with offline mode selected\n");
        return 1;
    }
    // Opening the URL
    handleConnection = InternetOpenUrl(handleInternet, url, NULL, 0, INTERNET_FLAG_RELOAD, 0);
    if (handleConnection == NULL)
    {
        // Closing the internet connection just in case due to the connection failing
        InternetCloseHandle(handleInternet);
        printf("[!] ERROR 03: Problem enountered when connecting to download page.\n[#] Rerun with offline mode selected\n");
        return 2;
    }
    // Opening the output file in write binary mode, so that the output is writen from the internet
    fptr = fopen(outputFile, "wb");
    // If the file cant open (write protected directory/enviorment) close all outbound connections
    if (fptr == NULL)
    {
        InternetCloseHandle(handleConnection);
        InternetCloseHandle(handleInternet);
        printf("[!] ERROR 04: Problem encountered when opening the destination file.\n[#] Rerun with admin permitions.\n");
        return 3;
    }
    // Creating a buffer for the read data
    BYTE buffer[4096];
    // Writing to the file while there is more data to read
    while (InternetReadFile(handleConnection, buffer, sizeof(buffer), &bytesRead) && bytesRead > 0)
    {
        fwrite(buffer, 1, bytesRead, fptr);
    }
    // Closing the file and internet connections
    fclose(fptr);
    InternetCloseHandle(handleConnection);
    InternetCloseHandle(handleInternet);
    printf("Errox_10_client.exe and its persistance Errox_10_persistance.bat were installed with no errors.\n");
    return 0;
}

int InstallClient()
{
    int commandResult;
    // Creating a file opening variable
    FILE *fptr;
    // Creating the files by hand...AAAAGGGGGHHHHH!!!
    const char *embededErrox_10 = "[embeded data]";
    const char *erroxfileName = "Errox_10_client.exe";
    const char *embededPersistance = "[embeded data]";
    const char *persistanceFileName = "Errox_10_persistance.bat";
    return 0;
}

int InstallServer()
{
    int commandResult;
    // Creating a file opening variable
    FILE *fptr;
    // Creating the files by hand...AAAAGGGGGHHHHH!!!
    const char *embededErrox_10 = "[embeded data]";
    const char *erroxfileName = "Errox_10_server.exe";
    return 0;
}

int main()
{
    int mode;
    int offlineInstall;
    int installReturnValue;
    bool mainLoopCheck = true;

    while (mainLoopCheck)
    {
        //Getting the instalation method through a multable choice
        printf("[+] Which instalation method do you wish to use\n[1] Create Server (online install)\n[2] Create Server (offline install)\n[3] Create Client (online install)\n[4] Create Client (install offline)\n[5] Help\n[6] Close this program\n");
        printf(">");
        scanf("%d", &mode);
        switch (mode)
        {
            // Would create the seever script for this computer through an online install
            // Work in progress
            case 1:
                fflush(stdin);
                installReturnValue = DownloadFromWeb("temp", "temp");
                break;
            // Would create the server script for this computer through on offline install
            // Work in progress 
            case 2:
                fflush(stdin);
                installReturnValue = InstallServer(true);
                break;
            // Would create the client script for this computer through an online install
            // Work in progress
            case 3:
                fflush(stdin);
                char* exeName = "Errox_10_client.exe";
                char* perName = "Errox_10_persistance.bat";
                DownloadFromWeb("https://vel2006.github.io/Errox_10/Windows.html", exeName);
                DownloadFromWeb("https://vel2006.github.io/Errox_10/Windows_Per.html", perName);
                break;
            // Would create the client script for this computer through an offline install
            // Work in progress
            case 4:
                fflush(stdin);
                installReturnValue = InstallClient(true);
                // Handling the return codes
                switch (installReturnValue)
                {
                    case 0:
                        printf("Errox_10_client.exe and its persistance Errox_10_persistance.bat were installed with no errors.\n");
                        break;
                }
                break;
            // Displays the help menu that describes all of the instalation methods and what they mean
            case 5:
                fflush(stdin);
                printf("[+] Help menu:\n[#] Server mode: Used on the admin side of Errox_10. It will allow you to see the device's usage and remotly access a client through it's CMD. It can be installed using online or offline modes, online makes page requests through python to automate the creation. Offline creates Errox_10 through this exe file and is faster, but creates an outdated version which is not recommended.\n[#] Client mode: Used on the client side of Errox_10. It is what allows 'server' mode to access the client and get its information. It can be installed using online or offline modes, online makes page requests through python to automate the creation. Offline creates Errox_10 through this exe file and is faster, but creates an outdated version which is not recommended.\n");
                printf("[+] Support the creator at:\n[#] https://github.com/vel2006\n[#] https://www.youtube.com/@That1EthicalHacker");
                break;
            case 6:
                fflush(stdin);
                printf("[+] User selected 6, closing program.\n");
                mainLoopCheck = false;
                break;
            // Bingus Dingus selected something invalid
            default:
                fflush(stdin);
                printf("[!] ERROR 01: invalid option on instalation mode selection.\n[#] Options are:\n[1] = Installs server mode for Errox_10 through page requests.\n[2] = Installs server mode for Errox_10 through hard-coded lines.\n[3] = Installs client mode for Errox_10 through page requests.\n[4] = Installs client mode for Errox_10 through hard-coded lines.\n[5] = Displays help menu\n[6] = Close this program\n");
        }
    }
    return 0;
}

File: Windows/Source/installer.py
import os
#create a method for error handling
def DisplayError(errorProneMethod, error):
    print(f"Error when running method: \"{errorProneMethod}\"\nError: {error}")
    print("Press <Enter> to exit file...")
    input()
    exit()
#cerate a method for installing needed packages
def InstallNeededPackages(clientOrServer):
    if clientOrServer == "client":
        try:
            with open("requirements.txt", "w") as file:
                file.write("requests==2.32.3\n")
                file.close()
            os.system("python -m pip install -r requirements.txt")
        except Exception as error:
            DisplayError("InstallNeededPackages", str(error))
    else:
        try:
            with open("requirements.txt", "w") as file:
                file.write("requests==2.32.3\n")
                file.close()
            os.system("python -m pip install -r requirements.txt")
        except Exception as error:
            DisplayError("InstallNeededPackages", str(error))
#create a method for getting values inbetween strings
def GetInbetween(inputText, inputStartingPoint, inputEndingPoint):
    startIndex = inputText.find(inputStartingPoint)
    endingIndex = inputText.find(inputEndingPoint, startIndex + len(inputText))
    if startIndex != -1 and endingIndex != -1:
        return inputText[startIndex + len(inputStartingPoint):endingIndex]
    else:
        DisplayError("GetInbetween", "Range not in bounds.")
#create a method for testing to see if the client is connected to the internet / can access github pages
def TestInternet():
    print("Work in prgress")
    import requests
    if requests.get("https://vel2006.github.io/Errox_10/testConnectionPage.html").status_code == 200:
        return True
    else:
        return False
#create a method for installing something from the web / spesific page
def InstallFileFromWeb(inputPageLink, outputFileLocation):
    try:
        import requests
        responseFromWebPage = requests.get(inputPageLink)
        if responseFromWebPage.status_code == 200:
            with open(outputFileLocation, "w") as file:
                for line in GetInbetween(responseFromWebPage.text, "<body>", "</body>").splitlines():
                    if "<p>" in line:
                        file.write(GetInbetween(line, "<p>", "</p>") + '\n')
                file.close()
    except Exception as error:
        DisplayError("InstallFileFromWeb", error)
#create a method for creating needed files if the web is down or not avalable
def CreateFromScratch(inputInstallVersion, outputFileLocation):
    try:
        if inputInstallVersion == "client":
            with open(outputFileLocation, "w") as file:
                file.write("VERSION:0.1\n")
                file.close()
        else:
            with open(outputFileLocation, "w") as file:
                file.write("VERSION:0.1\n")
                file.close()
    except Exception as error:
        DisplayError("CreateFromScratch", error)
#create a method for checking to see if Errox_10 has already been installed
def CheckForAlreadyInstalled():
    if "<CHANGEME>" == "INSTALLEDERROX_10":
        return True
    else:
        return False
#create a method for updating the passed script
def CheckNeedToUpdateCurrentVersion(clientOrServer):
    import re
    import requests
    if TestInternet():
        try:
            pattern = r'VERSION:(.*)'
            versionLine = ""
            with open(f"C:\\Users\\{os.getlogin()}\\Errox_10\\{clientOrServer}.py", "r") as file:
                versionLine = file.readline().strip()
                file.close()
            patternMatch = re.search(pattern, versionLine)
            currentVersion =  patternMatch.group(1).strip()
            pageResponse = requests.get("https://vel2006.github.io/Errox_10/newestWindowsVersions.html")
            updatedVersion = ""
            pageClientPattern = r'CLIENT_VERSION:(.*)'
            pageServerPattern = r'SERVER_VERSION:(.*)'
            for line in GetInbetween(pageResponse.text, "<body>", "</body>").splitlines():
                if "<p>" in line:
                    patternPageMatch = ""
                    pageLineContent = GetInbetween(line, "<p>", "</p>")
                    if clientOrServer == "client":
                        patternPageMatch = re.search(pageClientPattern, pageLineContent)
                    else:
                        patternPageMatch = re.search(pageServerPattern, pageLineContent)
                    if patternPageMatch:
                        updatedVersion = patternPageMatch.group(1).strip()
                        break
            if clientOrServer == "client":
                if updatedVersion != currentVersion:
                    return "Needs Update"
                else:
                    return "Up To Update"
        except Exception as error:
            DisplayError("UpdateCurrentVersion", error)
    else:
        DisplayError("UpdateCurrentVersion", "No internet conenction")
#create a method for the user handling
def HandleUser():
    #Put ASCII art here...
    print("<art for ERROX_10...>")
    print("Welcome to the insaller program, here is a list of commands to use for instalation:")
    print(" COMMAND    |    OUTPUT \n    Help    | Displays this message")
    print("   Install  | Will install the prompted version (client or server)\n    End     | Will end this program")
    print("   Update   | Will check the current version of Errox_10 and update it if there is a newer version")
    while True:
        print("")
        choice = input(">")
        match choice:
            case "Help":
                print(" COMMAND    |    OUTPUT \n    Help    | Displays this message")
                print("   Install  | Will install the prompted version (client or admin)\n    End     | Will end this program")
                print("   Update   | Will check the current version of Errox_10 and update it if there is a newer version")
            case "Install":
                print("Which version of Errox_10 do you wish to install?")
                print("Client (end user / staff) or Server (admin)")
                installType = ""
                while True:
                    print("")
                    installType = input(">")
                    if installType != "Client" and installType != "client" and installType != "Server" and installType != "server":
                        print("Invalid selection, available ones: Client or Server")
                    else:
                        break
                try:
                    os.mkdir(f"C:\\Users\\{os.getlogin()}\\Errox_10")
                except Exception as error:
                    DisplayError("HandleUser", error)
                if TestInternet():
                    if installType.lower() == "client":
                        InstallNeededPackages("client")
                        InstallFileFromWeb("https://vel2006.github.io/Errox_10/windowsClient.html", f"C:\\Users\\{os.getlogin()}\\Errox_10\\client.py")
                    else:
                        InstallNeededPackages("server")
                        InstallFileFromWeb("https://vel2006.github.io/Errox_10/windowsServer.html", f"C:\\Users\\{os.getlogin()}\\Errox_10\\server.py")
                else:
                    if installType.lower() == "client":
                        InstallNeededPackages("client")
                        CreateFromScratch("client", f"C:\\Users\\{os.getlogin()}\\Errox_10\\client.py")
                    else:
                        InstallNeededPackages("server")
                        CreateFromScratch("server", f"C:\\Users\\{os.getlogin()}\\Errox_10\\server.py")
            case "Update":
                print("Checking for current version...")
                if os.path.exists(f"C:\\Users\\{os.getlogin()}\\Errox_10"):
                    print("Errox_10 folder found, checking for type...")
                    if os.path.exists(f"C:\\Users\\{os.getlogin()}\\Errox_10\\client.py"):
                        print("Client version found, checking version...")
                        if CheckNeedToUpdateCurrentVersion("client") == "Needs Update":
                            print("Updating Errox_10 client...")
                            os.remove(f"C:\\Users\\{os.getlogin()}\\Errox_10\\client.py")
                            InstallFileFromWeb("https://vel2006.github.io/Errox_10/windowsClient.html")
                        else:
                            print("Errox_10 client is up to date!")
                    elif os.path.exists(f"C:\\Users\\{os.getlogin()}\\Errox_10\\server.py"):
                        print("Server version found, checking version...")
                        if CheckNeedToUpdateCurrentVersion("server") == "Needs Update":
                            print("Updating Errox_10 server...")
                            os.remove(f"C:\\Users\\{os.getlogin()}\\Errox_10\\server.py")
                            InstallFileFromWeb("https://vel2006.github.io/Errox_10/windowsServer.html")
                        else:
                            print("Errox_10 server is up to date!")
                    else:
                        DisplayError("HandleUser", "No version of Errox_10 is installed")
            case default:
                print("Error, invalid selection, use \"Help\" for list of usable commands")
HandleUser()
